File: nodegs/nodegsclient/nodegsclient/Service/EventService.cpp
#include "..\\stdafx.h"
#include "..\\CStr.h"
#include "EventService.h"
#include "Binary.h"

namespace OwLibCT
{
	EventService::EventService()
	{
		m_native = new NativeBase();
		m_socketID = 0;
		SetServiceID(SERVICEID_EVENT);
	}

	EventService::~EventService()
	{
		if(m_native)
		{
			delete m_native;
		}
	}

	NativeBase* EventService::GetNative()
	{
		return m_native;
	}

	int EventService::GetSocketID()
	{
		return m_socketID;
	}

	void EventService::SetSocketID(int socketID)
	{
		m_socketID = socketID;
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	int EventService::SendEvent(int requestID, GridA *grid)
	{
        int ret = Send(FUNCTIONID_EXCUTEID, requestID, grid);
        return ret > 0 ? 1 : 0;
	}

	GridA* EventService::GetGrid(CMessage *message)
	{
		Binary br;
		int size = message->m_bodyLength;
		br.Write(message->m_body, size);
		GridA *grid = new GridA();
		grid->SetNative(m_native);
		String name = L"";
		br.ReadString(name);
		grid->SetName(name);

		int columnsSize = br.ReadInt();
		for (int i = 0; i < columnsSize; i++)
		{
			GridColumn *column = new GridColumn();
			String name = L"";
			String columnType = L"";
			br.ReadString(name);
            br.ReadString(columnType);
			column->SetName(name);
			column->SetColumnType(columnType);
			grid->AddColumn(column);
		}
		grid->Update();
		vector<GridColumn*> columns = grid->GetColumns();
		int rowsCount = br.ReadInt();
		for (int i = 0; i < rowsCount; i++)
		{
			GridRow *row = new GridRow();
			grid->AddRow(row);
			for (int j = 0; j < columnsSize; j++)
			{
				GridColumn *column = columns[j];
				String columnType = column->GetColumnType();
				GridCell *cell;
				String value = L"";
				if (columnType == L"bool")
				{
					cell = new GridBoolCell();
					row->AddCell(j, cell);
					char charVal = br.ReadChar();
					if(charVal == '0')
					{
						cell->SetBool(false);
					}
					else
					{
						cell->SetBool(true);
					}
				}
				else if (columnType == L"double")
				{
					cell = new GridDoubleCell();
					row->AddCell(j, cell);
					cell->SetDouble(br.ReadDouble());
				}
				else if (columnType == L"float")
				{
					cell = new GridFloatCell();
					row->AddCell(j, cell);
					cell->SetFloat(br.ReadFloat());
				}
				else if (columnType == L"int")
				{
					cell = new GridIntCell();
					row->AddCell(j, cell);
					cell->SetInt(br.ReadInt());
				}
				else if (columnType == L"long")
				{
					cell = new GridLongCell();
					row->AddCell(j, cell);
					cell->SetLong((long)br.ReadDouble());
				}
				else if (columnType == L"string")
				{
					cell = new GridStringCell();
					row->AddCell(j, cell);
					br.ReadString(value);
					cell->SetString(value);
				}
				else
				{
					cell = new GridStringCell();
					row->AddCell(j, cell);
					br.ReadString(value);
					cell->SetString(value);
				}
			}
		}
		return grid;
	}

	
	void EventService::OnReceive(CMessage *message)
	{
		BaseService::OnReceive(message);
        SendToListener(message);
	}

	int EventService:: Send(int functionID, int requestID, GridA *grid)
	{
		int len = 0;
		Binary br;
		br.WriteString(grid->GetName());
		len += CStrA::GetWStringCapacity(grid->GetName());
        vector<GridColumn*> columns = grid->GetColumns();
		int columnsSize = (int)columns.size();
        br.WriteInt(columnsSize);
		len += 4;
        for (int i = 0; i < columnsSize; i++)
        {
            GridColumn *column = columns[i];
            br.WriteString(column->GetName());
			len += CStrA::GetWStringCapacity(column->GetName());
            br.WriteString(column->GetColumnType());
			len += CStrA::GetWStringCapacity(column->GetColumnType());
        }
        vector<GridRow*> rows = grid->GetRows();
		int rowsCount = (int)rows.size();
        br.WriteInt(rowsCount);
		len += 4;
        for (int i = 0; i < rowsCount; i++)
        {
            GridRow *row = rows[i];
            for (int j = 0; j < columnsSize; j++)
            {
                GridColumn *column = columns[j];
				String columnType = column->GetColumnType();
                GridCell *cell = row->GetCell(j);
                if (columnType == L"bool")
                {
					bool bolValue = cell->GetBool();
					if(!bolValue)
					{
						br.WriteChar(char(0));
					}else
					{
						br.WriteChar(char(1));
					}
					len +=1;
                }
                else if (columnType == L"double")
                {
                    br.WriteDouble(cell->GetDouble());
					len += 8;
                }
                else if (columnType == L"float")
                {
                    br.WriteFloat(cell->GetFloat());
					len += 4;
                }
                else if (columnType == L"int")
                {
                    br.WriteInt(cell->GetInt());
					len += 4;
                }
                else if (columnType == L"long")
                {
                    br.WriteDouble((double)cell->GetLong());
					len += 8;
                }
                else if (columnType == L"string")
                {
                    br.WriteString(cell->GetString());
					len += CStrA::GetWStringCapacity(cell->GetString());
                }
            }
        }
		char *bytes = (char*)br.GetBytes();

		CMessage message(GetGroupID(), GetServiceID(), functionID, GetSessionID(), requestID, m_socketID, 0, GetCompressType(), len, bytes);
		return BaseService::Send(&message);
	}
}

File: nodegs_cs/node.gs/Service/EventService.cs
using System;
using System.Collections.Generic;
using System.Text;
using OwLibSV;
using OwLib;

namespace node.gs
{
    /// <summary>
    /// 事件服务
    /// </summary>
    public class EventService:BaseService
    {
        /// <summary>
        /// 创建事件服务
        /// </summary>
        public EventService()
        {
            ServiceID = SERVICEID_EVENT;
        }

        private INativeBase m_native = NativeHandler.CreateNative();

        /// <summary>
        /// 事件服务ID
        /// </summary>
        public const int SERVICEID_EVENT = 30;

        /// <summary>
        /// 方法ID
        /// </summary>
        public const int FUNCTIONID_EXCUTEID = 0;

        private int m_socketID = 0;

        /// <summary>
        /// 获取或设置套接字ID
        /// </summary>
        public int SocketID
        {
            get { return m_socketID; }
            set { m_socketID = value; }
        }

        /// <summary>
        /// 获取表格数据
        /// </summary>
        /// <param name="grid">表格</param>
        /// <returns>流</returns>
        public byte[] GetBytes(GridA grid)
        {
            Binary br = new Binary();
            br.WriteString(grid.Name);
            List<GridColumn> columns = grid.GetColumns();
            int columnsSize = columns.Count;
            br.WriteInt(columnsSize);
            for (int i = 0; i < columnsSize; i++)
            {
                GridColumn column = columns[i];
                br.WriteString(column.Name);
                br.WriteString(column.ColumnType);
            }
            List<GridRow> rows = grid.GetRows();
            int rowsCount = rows.Count;
            br.WriteInt(rowsCount);
            for (int i = 0; i < rowsCount; i++)
            {
                GridRow row = rows[i];
                for (int j = 0; j < columnsSize; j++)
                {
                    GridColumn column = columns[j];
                    String columnType = column.ColumnType.ToLower();
                    GridCell cell = row.GetCell(j);
                    if (columnType == "bool")
                    {
                        br.WriteBool(cell.GetBool());
                    }
                    else if (columnType == "double")
                    {
                        br.WriteDouble(cell.GetDouble());
                    }
                    else if (columnType == "float")
                    {
                        br.WriteFloat(cell.GetFloat());
                    }
                    else if (columnType == "int")
                    {
                        br.WriteInt(cell.GetInt());
                    }
                    else if (columnType == "long")
                    {
                        br.WriteDouble(cell.GetLong());
                    }
                    else if (columnType == "string")
                    {
                        br.WriteString(cell.GetString());
                    }
                }
            }
            byte[] bytes = br.GetBytes();
            br.Close();
            return bytes;
        }

        /// <summary>
        /// 获取表格
        /// </summary>
        /// <param name="message"></param>
        /// <returns></returns>
        public int Excute(CMessage message)
        {
            GridA grid = GetGrid(message.m_body, m_native);
            Send(message, grid);
            return 1;
        }

        /// <summary>
        /// 获取表格
        /// </summary>
        /// <param name="bytes">流</param>
        /// <param name="native">方法库</param>
        /// <returns>表格</returns>
        public GridA GetGrid(byte[] bytes, INativeBase native)
        {
            Binary br = new Binary();
            br.Write(bytes, bytes.Length);
            GridA grid = new GridA();
            grid.Native = native;
            grid.Name = br.ReadString();
            int columnsSize = br.ReadInt();
            for (int i = 0; i < columnsSize; i++)
            {
                GridColumn column = new GridColumn();
                column.Name = br.ReadString();
                column.ColumnType = br.ReadString();
                grid.AddColumn(column);
            }
            grid.Update();
            List<GridColumn> columns = grid.GetColumns();
            int rowsCount = br.ReadInt();
            for (int i = 0; i < rowsCount; i++)
            {
                GridRow row = new GridRow();
                grid.AddRow(row);
                for (int j = 0; j < columnsSize; j++)
                {
                    GridColumn column = columns[j];
                    string columnType = column.ColumnType.ToLower();
                    GridCell cell = null;
                    if (columnType == "bool")
                    {
                        cell = new GridBoolCell();
                        row.AddCell(j, cell);
                        cell.SetBool(br.ReadBool());
                    }
                    else if (columnType == "double")
                    {
                        cell = new GridDoubleCell();
                        row.AddCell(j, cell);
                        cell.SetDouble(br.ReadDouble());
                    }
                    else if (columnType == "float")
                    {
                        cell = new GridFloatCell();
                        row.AddCell(j, cell);
                        cell.SetFloat(br.ReadFloat());
                    }
                    else if (columnType == "int")
                    {
                        cell = new GridIntCell();
                        row.AddCell(j, cell);
                        cell.SetInt(br.ReadInt());
                    }
                    else if (columnType == "long")
                    {
                        cell = new GridLongCell();
                        row.AddCell(j, cell);
                        cell.SetLong((long)br.ReadDouble());
                    }
                    else if (columnType == "string")
                    {
                        cell = new GridStringCell();
                        row.AddCell(j, cell);
                        cell.SetString(br.ReadString());
                    }
                    else
                    {
                        cell = new GridStringCell();
                        row.AddCell(j, cell);
                        cell.SetString(br.ReadString());
                    }
                }
            }
            br.Close();
            return grid;
        }

        /// <summary>
        /// 接收数据
        /// </summary>
        /// <param name="message">消息</param>
        public override void OnReceive(CMessage message)
        {
            base.OnReceive(message);
            switch (message.m_functionID)
            {
                case FUNCTIONID_EXCUTEID:
                    Excute(message);
                    break;
            }
        }

        /// <summary>
        /// 发送消息
        /// </summary>
        /// <param name="message">消息</param>
        /// <param name="grid">集合</param>
        public int Send(CMessage message, GridA grid)
        {
            byte[] bytes = GetBytes(grid);
            message.m_bodyLength = bytes.Length;
            message.m_body = bytes;
            int ret = Send(message);
            return ret;
        }
    }
}
